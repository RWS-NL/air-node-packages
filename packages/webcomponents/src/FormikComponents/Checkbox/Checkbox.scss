@import "../../styles/vars";

.checkBox {
  fill: rgba(0, 0, 0, 0.23);
}

.label {
  font-size: $fontSizeDefault;
}
