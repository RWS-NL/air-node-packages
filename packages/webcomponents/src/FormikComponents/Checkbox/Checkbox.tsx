import { Checkbox as MUICheckbox, FormControlLabel } from '@material-ui/core';
import CheckBoxOutlineFilledIcon from '@material-ui/icons/CheckBox';
import CheckBoxOutlineBlankIcon from '@material-ui/icons/CheckBoxOutlineBlank';
import React, { ChangeEvent, forwardRef, ForwardRefExoticComponent } from 'react';
import css from './Checkbox.scss';

export interface CheckboxProps {
  /** The name for the checkbox field */
  name: string;
  /** A tracker for the current value of the checkbox */
  value: string;
  /** Data-QA to send to the checkbox for testing */
  'data-qa'?: string;
  /**
   * The formik onChange event
   * @example formikProps.handlechange
   */
  onChange: (e: ChangeEvent<any>) => void;
  /**
   * The formik onBlur event
   * @example formikProps.handleBlur
   */
  onBlur: (...args: unknown[]) => void;
}

/**
 * Constructs a Checkbox with formik validation
 * @param props Props to pass to the Checkbox component
 * @example
 * <FieldArray
 *     name='fieldArray'
 *     render={arrayHelpers => (
 *     <Fragment>
 *         {arrayWithObjects.map((item, index) => (
 *         <Grid item xs={4} key={index}>
 *             <Tooltip title={`form.title.${item.name.toLowerCase()}`} placement='top'>
 *             <Checkbox
 *                 key={index}
 *                 value={item.name}
 *                 data-qa={`checkbox-${item.name.toLowerCase()}`}
 *                 name={`form.title.${item.name.toLowerCase()}`}
 *                 onChange={(e: React.ChangeEvent<{ checked: boolean }>) => {
 *                 if (e.target.checked) {
 *                     arrayHelpers.push(item.name);
 *                 } else {
 *                     const idx = formikProps.values.fieldName.indexOf(item.name);
 *                     arrayHelpers.remove(idx);
 *                 }
 *                 }}
 *                 onBlur={formikProps.handleBlur}
 *             />
 *             </Tooltip>
 *         </Grid>
 *         ))}
 *     </Fragment>
 *     )}
 * />
 */
export const Checkbox: ForwardRefExoticComponent<CheckboxProps> = forwardRef(({
  name, value, onChange, onBlur, 'data-qa': dataQa, ...props
}, ref) => (
  <FormControlLabel
    {...props}
    ref={ref}
    control={
      <MUICheckbox
        icon={<CheckBoxOutlineBlankIcon className={css.checkBox} fontSize='small' />}
        checkedIcon={<CheckBoxOutlineFilledIcon fontSize='small' />}
        name={name}
        value={value}
        onChange={onChange}
        onBlur={onBlur}
        color='primary'
        data-qa={dataQa}
      />
    }
    label={name}
    classes={{ label: css.label }}
  />
));

export default Checkbox;