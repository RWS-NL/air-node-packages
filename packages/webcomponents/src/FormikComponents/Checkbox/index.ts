export { default as Checkbox, CheckboxProps } from './Checkbox';