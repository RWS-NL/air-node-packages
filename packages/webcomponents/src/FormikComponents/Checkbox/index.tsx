import MUICheckbox, { CheckboxProps as MUICheckboxProps } from '@mui/material/Checkbox';
import FormControlLabel, { FormControlLabelProps } from '@mui/material/FormControlLabel';
import CheckBoxOutlineFilledIcon from '@mui/icons-material/CheckBox';
import CheckBoxOutlineBlankIcon from '@mui/icons-material/CheckBoxOutlineBlank';
import { Field, FieldProps } from 'formik';
import React, { ChangeEvent, forwardRef } from 'react';
import { Tooltip, TooltipProps } from '../../Tooltip';
import css from './Checkbox.scss';

/**
 * Properties that can be passed to the Checkbox component
 * @extends [Checkbox API](https://material-ui.com/api/checkbox/#checkbox-api)
 */
export interface CheckboxProps extends MUICheckboxProps {
  /**
   * The name for this specific checkbox
   *
   * @remarks
   * It is recommend that this is different from the name given to the encapsulating FieldArray
   */
  name: string;
  /** The label to display next to the checkbox */
  label: string;
  /** A tracker for the current value of the checkbox */
  value: string;
  /** The text to display inside the Tooltip that pops up when hovering over the checkbox or the label */
  tooltipText: string;
  /**
   * Whether the checkbox should be in a checked state
   * @default false
   */
  checked?: boolean;
  /** Data-QA to send to the checkbox for testing */
  'data-qa'?: string;

  /** Props applied to the [Tooltip](https://rws-nl.github.io/air-node-packages/modules/_rws_air_webcomponents.html#tooltip-1) component */
  TooltipProps?: TooltipProps;
  /** Props applied to the [FormControlLabel](https://material-ui.com/api/form-control-label/#formcontrollabel-api) component */
  FormControlLabelProps?: FormControlLabelProps;
}

export type CheckboxChangeEvent = ChangeEvent<{ checked: boolean }>;

/**
 * Constructs a Checkbox with formik validation
 * @param props Props to pass to the Checkbox component
 * @example
 * ```jsx
 *  <FieldArray
 *    name='fieldArrayName'
 *    render={arrayHelpers =>
 *      arrayWithStrings((entry, index) => (
 *        <Grid item xs={4} key={index}>
 *          <Checkbox
 *            name={entry}
 *            value={entry}
 *            label={t(`form.fieldArrayName.labels.${entry}`)}
 *            tooltipText={t(`form.fieldArrayName.tooltips.${entry}`)}
 *            data-qa={`checkbox-${entry.toLowerCase()}`}
 *            checked={values['fieldArrayName'].includes(entry)}
 *            onChange={(e: CheckboxChangeEvent) => {
 *              if (e.target.checked) {
 *                arrayHelpers.push(entry);
 *              } else {
 *                const idx = values['fieldArrayName'].indexOf(entry);
 *                arrayHelpers.remove(idx);
 *              }
 *            }}
 *          />
 *        </Grid>
 *      ))
 *    }
 *  />
 * ```
 */
export const Checkbox = forwardRef<HTMLButtonElement, CheckboxProps>(
  (
    { label, value, name, onChange, tooltipText, TooltipProps, FormControlLabelProps, checked = false, ...props },
    ref
  ) => (
    <Field type='checkbox'>
      {({ field }: FieldProps<string>) => (
        <Tooltip title={tooltipText} placement='top' {...TooltipProps}>
          <FormControlLabel
            {...field}
            ref={ref}
            label={label}
            classes={{ label: css.label }}
            control={
              <MUICheckbox
                {...props}
                color='primary'
                checkedIcon={<CheckBoxOutlineFilledIcon fontSize='small' />}
                icon={<CheckBoxOutlineBlankIcon className={css.checkBox} fontSize='small' />}
                value={value}
                name={name}
                checked={checked}
                onChange={onChange}
              />
            }
            {...FormControlLabelProps}
          />
        </Tooltip>
      )}
    </Field>
  )
);
