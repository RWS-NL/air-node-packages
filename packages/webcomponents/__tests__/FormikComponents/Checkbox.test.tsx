import Checkbox from '../../src/FormikComponents/Checkbox';
import { shallow, ShallowWrapper } from 'enzyme';
import MUICheckbox from '@material-ui/core/Checkbox/Checkbox';
import React from 'react';

const mockCheckboxOnChange = jest.fn();
const mockCheckboxOnBlur = jest.fn();

describe('Component Tests', () => {
  let checkBox: ShallowWrapper;

  beforeAll(() => {
    checkBox = shallow(
      <Checkbox name='testCheckbox' value='testCheckbox' onBlur={mockCheckboxOnBlur} onChange={mockCheckboxOnChange} />
    );
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  test('checkbox function called onChange', () => {
    const changeEvent = { event: { target: { value: true } } };
    checkBox
      .dive()
      .dive()
      .find(MUICheckbox)
      .simulate('change', changeEvent);
    expect(mockCheckboxOnChange).toHaveBeenCalledWith(changeEvent);
    expect(mockCheckboxOnChange).toHaveBeenCalledTimes(1);
  });
});

describe('Snapshot Testing', () => {
  test('Required Props', () => {
    const checkBox = shallow(
      <Checkbox name='testCheckbox' value='testCheckbox' onBlur={mockCheckboxOnBlur} onChange={mockCheckboxOnChange} />
    );
    expect(checkBox).toMatchSnapshot();
  });
});
