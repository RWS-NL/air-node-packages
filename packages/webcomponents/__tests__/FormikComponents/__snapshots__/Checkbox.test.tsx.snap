// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing Required Props 1`] = `
<WithStyles(ForwardRef(FormControlLabel))
  classes={
    Object {
      "label": "label",
    }
  }
  control={
    <WithStyles(ForwardRef(Checkbox))
      checkedIcon={
        <Memo(CheckBoxIcon)
          fontSize="small"
        />
      }
      color="primary"
      icon={
        <Memo(CheckBoxOutlineBlankIcon)
          className="checkBox"
          fontSize="small"
        />
      }
      name="testCheckbox"
      onBlur={[MockFunction]}
      onChange={[MockFunction]}
      value="testCheckbox"
    />
  }
  label="testCheckbox"
/>
`;
